core/node/node_storage_init/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+use std::{future::Future, sync::Arc, time::Duration};
+
+use tokio::sync::watch;
+use zksync_config::ObjectStoreConfig;
+use zksync_dal::{ConnectionPool, Core, CoreDal as _};
+use zksync_types::L1BatchNumber;
+
+pub use crate::traits::{InitializeStorage, RevertStorage};
+
+pub mod external_node;
+pub mod main_node;
+mod traits;
+
+#[derive(Debug)]
+pub struct SnapshotRecoveryConfig {
+    /// If not specified, the latest snapshot will be used.
+    pub snapshot_l1_batch_override: Option<L1BatchNumber>,
+    pub drop_storage_key_preimages: bool,
+    pub object_store_config: Option<ObjectStoreConfig>,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum InitDecision {
+    /// Perform or check genesis.
+    Genesis,
+    /// Perform or check snapshot recovery.
+    SnapshotRecovery,
+}
+
+#[derive(Debug)]
+pub struct NodeInitializationStrategy {
+    pub genesis: Box<dyn InitializeStorage>,
+    pub snapshot_recovery: Option<Box<dyn InitializeStorage>>,
+    pub block_reverter: Option<Box<dyn RevertStorage>>,
```

Comment:
Nit: Maybe it makes sense to use `Arc<_>` for field to avoid wrapping strategy itself in an `Arc`?